Sources/Fuzzilli/FuzzIL/AbstractInterpreter.swift

```diff
@@ -35,83 +126,28 @@ public struct AbstractInterpreter {
     }
     
     public mutating func reset() {
-        ifStack.removeAll()
-        stack = [VariableMap<Type>()]
+        state.reset()
         propertyTypes.removeAll()
         methodSignatures.removeAll()
     }
 
-    /// Get variables with new type which have different type in old state
-    /// Used after block end when some state on stack is removed
-    private func getStateChanges(oldState: VariableMap<Type>, newState: VariableMap<Type>) -> [(Variable, Type)] {
-        var typeChanges = [(Variable, Type)]()
-        for (variable, type) in newState {
-            let oldType = oldState[variable]
-            if oldType == nil || type != oldType  {
-                typeChanges.append((variable, type))
-            }
-        }
-        return typeChanges
-    }
-
     /// Abstractly execute the given instruction, thus updating type information.
     /// Return type changes.
     public mutating func execute(_ instr: Instruction) -> [(Variable, Type)] {
         var typeChanges = [(Variable, Type)]()
         switch instr.op {
-        case is BeginAnyFunctionDefinition:
-            stack.append(currentState)
-        case is EndAnyFunctionDefinition:
-            let functionState = stack.removeLast()
-            let previousState = stack.removeLast()
-            stack.append(merge(functionState, previousState))
-            typeChanges += getStateChanges(oldState: functionState, newState: stack.last!)
         case is BeginIf:
-            stack.append(currentState)
+            state.pushNewState()
         case is BeginElse:
-            let ifState = stack.removeLast()
-            ifStack.append(ifState)
-            typeChanges += getStateChanges(oldState: ifState, newState: stack.last!)
+            state.saveIfState()
+            // Exactly one of if/else block is executed. So we should reset types from if block now.
```

Comment:
Maybe the comment should also mention that the else block is executed from the state *before* the if block, so that's why the reset is necessary?